src/combinatorics/ssyt.lean

```diff
@@ -0,0 +1,85 @@
+/-
+Copyright (c) 2022 Jake Levinson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jake Levinson
+-/
+import combinatorics.young_diagram
+
+/-!
+# Semistandard Young tableaux
+
+A semistandard Young tableau is a filling of a Young diagram by natural numbers, such that
+the entries are weakly increasing left-to-right along rows (i.e. for fixed `i`), and
+strictly-increasing top-to-bottom along columns (i.e. for fixed `j`).
+
+An example of an SSYT of shape `μ = [4, 2, 1]` is:
+
+```text
+0 0 0 2
+1 1
+2
+```
+
+We represent an SSYT as a function `ℕ → ℕ → ℕ`, which is required to be zero for all pairs
+`(i, j) ∉ μ` and to satisfy the row-weak and column-strict conditions on `μ`.
+
+
+## Main definitions
+
+- `ssyt (μ : young_diagram)` : semistandard Young tableaux of shape μ
+- `ssyt.highest_weight (μ : young_diagram)`: the semistandard Young tableau whose `i`th row
+  consists entirely of `i`s, for each `i`.
+
+## Tags
+
+Semistandard Young tableau
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+/-- A semistandard Young tableau (SSYT) is a filling of the cells of a Young diagram by natural
+numbers, such that the entries in each row are weakly increasing (left to right), and the entries
+in each column are strictly increasing (top to bottom).
+
+Here, an SSYT is represented as an unrestricted function `ℕ → ℕ → ℕ`, but which is required to
+vanish outside `μ`. --/
+@[ext] structure ssyt (μ : young_diagram) :=
+  (entry : ℕ → ℕ → ℕ)
+  (row_weak : ∀ {i j1 j2 : ℕ}, j1 < j2 → (i, j2) ∈ μ → entry i j1 ≤ entry i j2)
+  (col_strict : ∀ {i1 i2 j: ℕ}, i1 < i2 → (i2, j) ∈ μ → entry i1 j < entry i2 j)
+  (zeros' : ∀ {i j}, (i, j) ∉ μ → entry i j = 0)
```

Comment:
```suggestion
(entry : ℕ → ℕ → ℕ)
(row_weak : ∀ {i j1 j2 : ℕ}, j1 < j2 → (i, j2) ∈ μ → entry i j1 ≤ entry i j2)
(col_strict : ∀ {i1 i2 j: ℕ}, i1 < i2 → (i2, j) ∈ μ → entry i1 j < entry i2 j)
(zeros' : ∀ {i j}, (i, j) ∉ μ → entry i j = 0)
```

Design question: had you considered making `entry` be a function `ℕ × ℕ → ℕ` instead? For example, later on you can write `λ c, ite (c ∈ μ) c.1 0`. There's certainly a tradeoff, but it seems to me that this makes `entry` more directly be a function "on" the `young_diagram`.

Reply:
I did consider it. There are some other places in the library where various lemmas are written involving two natural numbers `(i j : ℕ)` that could conceivably be formulated for a pair `c : ℕ × ℕ`.

The reason I picked `ℕ → ℕ → ℕ` is because I expected to frequently encounter situations where one of the coordinates is easier to reason about than the other, and only rarely encounter bundled pairs `c : ℕ × ℕ`. For example, RSK row insertion will eventually be defined by recursively on pairs `i insertion_value : ℕ` (and with `j` then defined as a function of `i` and `insertion_value`).

That said, the `fintype` instance for SSYTs with fixed shape and bounded entries `< m` will be from the equivalence with the appropriate subtype of functions `μ.cells → fin m`. I didn't end up using this elsewhere, though.

Actually, there's one other design option here: we could drop the `ssyt.zeros'` condition and instead define `ssyt` as a quotient of the function type. I didn't do this because I'm not sufficiently familiar with quotients in Lean.

Reply:
This probably deserves a longer discussion though. I don't claim to be certain that this setup is the best way to proceed. I am pretty sure the function should be on unrestricted natural numbers, but that would be fine regardless of whether it's `ℕ → ℕ → ℕ` or `ℕ × ℕ → ℕ`.

Reply:
@kmill There is actually another, possibly more important design question here, which is: 
1) Should the `young_diagram` argument be bundled into the SSYT?
2) Should we instead just have `is_ssyt : young_diagram → (ℕ → ℕ → ℕ) → Prop` rather than a `structure`?

Reply:
Discussed this on Zulip [here](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2316195.20ssyts). Conclusion was, let's stick with `entry : ℕ → ℕ → ℕ` as is.

Reply:
@jakelev I'm not sure there was really much of a conclusion to the discussion (I've found that when I write questions as long as yours on Zulip it tends to get less engagement, but that's just what happens when it's a bunch of busy volunteers), but it's fine leaving it as `ℕ → ℕ → ℕ` though so long as you keep `ℕ × ℕ → ℕ` in mind in case it makes sense to change it at some point.

Bundling a SSYT seems like a good design. By the way, with the newest lean/mathlib, dot notation has been changed such that you have access to everything in the `young_diagram` namespace from an `ssyt`.

Regarding quotients, my rule of thumb is that if there is a type that satisfies the universal property of a quotient, use that instead of a `quot`/`quotient`. Here, there is an obvious representative we can use for each `ssyt` by setting values to `0` outside the diagram. It's a similar rule of thumb for subtypes: rather than `μ.cells → ℕ` just set things to a reasonable junk value outside the domain if possible, since then such functions satisfy the universal property for functions from a subtype.

Reply:
Also, please make sure you incorporate the suggested whitespace changes in the first comment.

Reply:
Will do! I'm happy to discuss the ssyt design choice further if you have time. I do think it is a significant choice!